### PR TITLE
[release 2.0] expression: refactor scalarFuncToPBExpr (#6290)

### PR DESCRIPTION
**Pull request by zz-jason**
cherry pick #6290 to release 2.0

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@zz-jason ci failed.

**Comment by zz-jason** [comment]
@XuHuaiyu blocked by https://github.com/pingcap/tidb/pull/6388

File: expression/expr_to_pb.go [hunk 1]
```diff
@@ -338,63 +248,54 @@ func SortByItemToPB(sc *stmtctx.StatementContext, client kv.Client, expr Express
...
+		ast.JSONType,
+		ast.JSONExtract,
+		ast.JSONUnquote,
+		ast.JSONObject,
+		ast.JSONArray,
+		ast.JSONMerge,
+		ast.JSONSet,
+		ast.JSONInsert,
+		ast.JSONReplace,
+		ast.JSONRemove,
```
Comment (coocood) [inline expression/expr_to_pb.go]: Missing `JSONContains` and `JSONValid`?

File: expression/expr_to_pb.go [hunk 2]
```diff
@@ -338,63 +248,54 @@ func SortByItemToPB(sc *stmtctx.StatementContext, client kv.Client, expr Express
...
-	if pc.client.IsRequestTypeSupported(kv.ReqTypeDAG, kv.ReqSubTypeSignature) {
-		code := expr.Function.PbCode()
-		if code > 0 {
-			return &tipb.Expr{Tp: tipb.ExprType_ScalarFunc, Sig: code, Children: children, FieldType: toPBFieldType(expr.RetType)}
-		}
-		return nil
+		// control flow functions.
+		ast.Case,
+		ast.If,
+		ast.Ifnull,
```
Comment (coocood) [inline expression/expr_to_pb.go]: NullIf is missing?

**Comment by coocood** [comment]
This commit is unneccesary to cherry-pick

**Pull request closed without merging**